File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/IDCardToolkitObjCApp/IDCardToolkitObjCApp/ViewController/ListReadersViewController.m
//
//  ListReadersViewController.m
//  IDCardToolkitObjCApp
//
//  Created by Federal Authority For Identity and Citizenship on 20/12/17.
//  Copyright © 2017 Federal Authority For Identity and Citizenship. All rights reserved.
//

#import "ListReadersViewController.h"

@interface ListReadersViewController () {
    
    NSMutableArray *readerArr;
    NSArray *readers;
}
@property(nonatomic,strong) CardReader *cardreader;
@property(nonatomic,strong)Utils * utils;
@end

@implementation ListReadersViewController
@synthesize cardreader;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title=@"List of Readers ";
    self.navigationItem.hidesBackButton = YES;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"backbutton"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal]style:UIBarButtonItemStylePlain target:self action:@selector(goBack)];
    self.listReaderButton.layer.cornerRadius=22.0;
    [self.listReaderButton.layer setShadowOffset:CGSizeMake(5, 5)];
    [self.listReaderButton.layer setShadowColor:[[UIColor lightGrayColor] CGColor]];
    [self.listReaderButton.layer setShadowOpacity:0.5];
//    self.listReaderButton.layer.borderWidth=1.0;
    
    readerArr=[[NSMutableArray alloc]init];
    self.listReaderPickerView.hidden=YES;
    
    AppDelegate *aDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    self.model = aDelegate.model;
    self.utils=aDelegate.utils;
    
    // Do any additional setup after loading the view.
}
-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:YES];
    
    [readerArr removeAllObjects];
}
- (IBAction)listReaderButtonAction:(id)sender {
    
    [readerArr removeAllObjects];
    
    [self.utils ShowProgressBar:LISTREADERS andView:self.view];
//    dispatch_async(dispatch_get_global_queue( DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^(void){
        @try {
            Toolkit *toolkit = self.model.toolkitShared;
            self->readers=[toolkit listReaders];
            NSLog(@"readers %@",self->readers);
        }//try
        @catch (NSException *exception) {
            NSString *exe =[NSString stringWithFormat:@"%@\n%@\n%@",exception.name,exception.description,exception.userInfo];
            [AlertView showAlertTitle:ALERT withMessage:exe onView:self];
        }// catch
        @finally {
            dispatch_async(dispatch_get_main_queue(), ^{
                for (int i =0; i<self->readers.count; i++) {
                    NSString *readersName = [(CardReader *)[self->readers objectAtIndex:i] getName];
                    [self->readerArr addObject:readersName];
                }//for
                if (self->readerArr.count>0) {
                    self.listReaderPickerView.hidden=NO;
                    [self.listReaderPickerView reloadAllComponents];
                    
                    self->cardreader=(CardReader *)[self->readers objectAtIndex:0];
                    [self.model.cardReaderSharedArray removeAllObjects];
                    [self.model.cardReaderSharedArray addObject:self->cardreader];
                }// if
                else {
                    self.listReaderPickerView.hidden=YES;
                    [self.listReaderPickerView reloadAllComponents];
                }// else
                [self.utils DismissProgressBar];
            });// update UI main queue
        }// finally
   // });// background queue
}
- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}
#pragma mark PickerView Delagates methods

- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}
- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return readerArr.count;
}
- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    if (pickerView == self.listReaderPickerView  && component==0)
        return [readerArr objectAtIndex:row];
    
    return nil;
}
- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component {
    
    cardreader=(CardReader *)[self->readers objectAtIndex:row];
    [self.listReaderPickerView selectRow:row inComponent:0 animated:NO];
    
    [self.model.cardReaderSharedArray removeAllObjects];
    [self.model.cardReaderSharedArray addObject:cardreader];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */


@end

File: id-card-toolkit-ios-sdk-v3.1.6/id-card-toolkit-ios-sdk-v3.1.6/samples/ICASwiftApp/ICASwiftApp/ListReadersViewController.swift
//
//  ListReadersViewController.swift
//  ICASwiftApp
//
//  Created by Digital Trust Tech on 04/05/22.
//

import UIKit
import CoreBluetooth

class ListReadersViewController: UIViewController,UIPickerViewDataSource, UIPickerViewDelegate {
    
    @IBOutlet var listReaderButton: UIButton!
    @IBOutlet var listReaderPickerView: UIPickerView!
    @IBOutlet weak var view_Roundedcorners: UIView!
    
    //    var  readers = NSArray()
    var readerArr: [Any] = []
    var readers: [Any] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        self.view_Roundedcorners.cornered_View([.layerMaxXMaxYCorner, .layerMinXMaxYCorner], radius: 20, borderColor: .red, borderWidth: 0)
        
        
        
        listReaderButton.layer.cornerRadius = 22.0
        listReaderButton.layer.shadowOffset = CGSize(width: 5.0, height: 5.0)
        listReaderButton.layer.shadowColor = UIColor.lightGray.cgColor
        listReaderButton.layer.shadowOpacity = 0.5
        listReaderPickerView.isHidden = true
    }
    
    @IBAction func btn_BackAction(_ sender: Any) {
        navigationController?.popViewController(animated: true)
    }
    
    @IBAction func listReaderButtonAction(_ sender: Any) {
        
        if toolkit == nil  {
            self.view.makeToast(Constants.toolkitNotInitilized)
            return
        }//if
        
        utils.showProgressBar(Constants.ListReader, andView: self.view)
        readerArr.removeAll()
        readers.removeAll()
        //        DispatchQueue.global(qos: .background).async {
        do {
            self.readers = try toolkit?.listReaders() as! [Any]
            DispatchQueue.main.async {
                for i in 0..<self.readers.count {
                    let readersName: String? = (self.readers[i] as? CardReader)?.getName()
                    self.readerArr.append("\(readersName!)")
                }//for
                if self.readerArr.count > 0 {
                    self.listReaderPickerView.isHidden = false
                    self.listReaderPickerView.reloadAllComponents()
                    cardreader = self.readers[0] as? CardReader
                }//if
                else {
                    AlertView.showAlertTitle(Constants.Alert, withMessage: "Readers Not Found", onView: self)
                }//else
                utils.dismissProgressBar()
            }// update UI main queue
        }//do
        catch let error as NSError{
            DispatchQueue.main.async {
                let err="\(error.domain)\n\(error.code)\n\(error.userInfo)"
                AlertView.showAlertTitle(Constants.Alert, withMessage:err,onView: self)
                utils.dismissProgressBar()
                self.listReaderPickerView.isHidden = true
                self.listReaderPickerView.reloadAllComponents()
            }// update UI main queue
        }//catch
        // }//background queue
    }
    // MARK:PickerView Delagates methods
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return readerArr.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        if pickerView == self.listReaderPickerView && component == 0 {
            return readerArr[row] as? String
        }// if
        return nil
    }
    
    func pickerView(_ pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        cardreader = self.readers[row] as? CardReader
        self.listReaderPickerView.selectRow(row, inComponent: 0, animated: true)
    }
}
